Doc reorg: Scratch Orgs (#2042)
SFDO-Tooling/CumulusCI: docs/scratch_orgs.rst

Manage Scratch Orgs
===================

> Throughout this section, we'll focus on scratch org management in the CumulusCI keychain. To learn about managing persistent orgs, such as sandboxes, production orgs, and packaging orgs, read TODO: reference "Connecting persistent orgs".

Reviewer: We should probably define what a scratch org is and say a bit about why it's our preferred way to do development.

Author: What do you think of the paragraphs added above? I don't want to go into too much detail here, thinking that deeper advocacy probably belongs in other sections, but tried to give a quick overview.

What is an Org in CumulusCI?
----------------------------

> CumulusCI takes a different approach to creating and using scratch orgs that aims to make the process easier, more portable, and more repeatable. An org in CumulusCI's keychain starts out as a named configuration, tailored for a specific purpose within the lifecycle of the project. The scratch org is only actually generated the first time you use the scratch org from the project keychain - and once it's expired or been deleted, a new one can easily be created with the same configuration.

Reviewer: > CumulusCI takes a different approach to creating and using scratch orgs that aims to make the process easier, more portable, and more repeatable.

Compared to what? It feels like you're assuming that the reader is familiar with using scratch orgs from sfdx but it's just as likely that the org development model in sandboxes is what they're familiar with.

CumulusCI offers tools that make it easy to discover predefined org configurations, create scratch orgs based on those configurations, and define new orgs and new configurations.

Set Up the Salesforce CLI
-------------------------

If you haven't already set up the Salesforce CLI, you need to take care of a few steps. For a detailed introduction to setting up Salesforce CLI and Visual Studio Code to work with CumulusCI, we recommend completing `Build Applications with CumulusCI <https://trailhead.salesforce.com/en/content/learn/trails/build-applications-with-cumulusci>`_ on Trailhead.

1. `Install the Salesforce CLI <https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_install_cli.htm>`_
2. `Enable Dev Hub in Your Org <https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_enable_devhub.htm>`_
3. `Connect SFDX to Your Dev Hub Org <https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_auth_web_flow.htm>`_ (be sure to use the ``--setdefaultdevhubusername`` option).

> If you already have the ``sfdx`` command installed, have connected to your Dev Hub, and have set the ``defaultdevhubusername`` config setting (use ``sfdx force:config:list`` to verify), you're ready to start using CumulusCI with Salesforce CLI.

Reviewer: Probably better to verify using `force:org:list` and looking for the `(D)` indicator. `force:config:list` can be confusing because of the distinction between global- and project-level settings. It's also [deprecated](https://github.com/forcedotcom/cli/blob/master/releasenotes/README.md#4980-august-27-2020)

> You can learn more about the Salesforce CLI at https://developer.salesforce.com/platform/dx.

Reviewer: We should probably standardize on a format similar to other Salesforce docs for linking to other resources: a "See Also" subheading with a list of links. This should probably go in our style guide too...

Predefined Orgs
---------------

CumulusCI comes with five org predefined org configurations. Every org's keychain starts with these configurations available and ready to be turned into a live scratch org:

* ``dev`` is for development workflows and uses the ``orgs/dev.json`` configuration file. It has a seven-day lifespan.
* ``qa`` is for testing workflows. ``qa`` and ``dev`` are the same out of the box, but can be customized to suit the needs of the project. It has a seven-day lifespan.
* ``feature`` is for use in continuous integration and uses the ``orgs/dev.json`` configuration file. It has a one-day lifespan.
* ``beta`` is for use in continuous integration or hands-on testing. It uses the ``orgs/beta.json`` configuration. It has a one-day lifespan.
* ``release`` is for use in continuous integration, hands-on testing, or demo workflow. It uses the ``orgs/release.json`` configuration. It has a one-day lifespan.

> 

Reviewer: Show `cci org list` here and point out that these are shown there before doing anything else?

Reviewer: Is there any compelling reason we have both dev and feature?

Author: I don't know of any reason for both of those shapes to exist.

Create a Scratch Org
--------------------

To create a scratch org from a configuration, simply use it as the target of a CumulusCI command, task, or flow. CumulusCI automatically initializes orgs when they're first used. To create a scratch org from the ``dev`` configuration and review information about the created org, run

.. code-block:: console

    $ cci org info dev

Once the org is created, it's associated with the name ``dev`` in the CumulusCI keychain and can be used with other commands until it expires. Once an org expires or is removed, its associated configuration is left in place, and can be recreated whenever needed.

> You can create new org configurations that inherit their configuration from a built-in org. For example, to create a new org name that uses the same configuration as type ``dev``, you can use the command ::

Reviewer: I'd be inclined to put this in a separate section further down after covering other more basic operations.

If you want to permanently remove an org from the org list, you can use ``cci org remove`` which will completely remove the org from the list.  If a scratch org has already been created from the config, the associated scratch org will also be deleted.

.. code-block:: console

    $ cci org remove feature-123

It's not necessary to explicitly remove or delete expired orgs. CumulusCI will recreate an expired org the first time you attempt to use it. To clean up expired orgs from the keychain, you can use the ``cci org prune`` command:

.. code-block:: console

    $ cci org prune

Set a Default Org
-----------------

When you run a Flow or Task that performs work on an org, you specify the org with a ``--org <name>`` option:

.. code-block:: console

    $ cci flow run dev_org --org dev

If you're running many commands against the same org, you may wish to set a default:

.. code-block:: console

    $ cci org default dev
    $ cci flow run dev_org

To remove an existing default, run the command

.. code-block:: console

    $ cci org default dev --unset

> 

Reviewer: An org can also be set as the default while adding it using `cci org scratch`

Manage Scratch Orgs
===================

*Scratch orgs* are ephemeral Salesforce orgs. Scratch orgs are created empty, unlike sandboxes, and configured by deploying metadata into them. Scratch orgs facilitate repeatable, fully source-driven development processes on the Salesforce platform, and help ameliorate many of the challenges inherent in working with *persistent orgs* like sandboxes and Developer Edition orgs. Scratch orgs are intended to be created and destroyed often; they cannot have a lifespan greater than 30 days.

> The CumulusCI Suite offers tools for working with all types of Salesforce orgs, but provides the most value when working with scratch orgs. We strongly encourage building processes around scratch orgs for development, feature and integration testing, demos, and other workflows. CumulusCI automation helps realize the promise of scratch orgs as low cost, repeatable, source-driven environments for every phase of the product lifecycle.

Reviewer: My take on the above paragraphs:

Scratch orgs are temporary Salesforce orgs which can be quickly set up "from scratch" and which last for a limited time (no more than 30 days). We strongly encourage using scratch orgs for development and QA, instead of sandboxes or Developer Edition orgs, because:

- They provide a repeatable starting point without the cruft that can accumulate in persistent orgs over time.
- The limited lifespan motivates making sure that all steps to set up the org are performed by CumulusCI automation with configuration stored as part of the project's source repository, which makes it easier to repeat the same steps in other orgs.

Throughout this section, we'll focus on managing scratch orgs in a CumulusCI project. To learn about managing persistent orgs, such as sandboxes, production orgs, and packaging orgs, read TODO: reference "Connecting persistent orgs".

What is an Org in CumulusCI?
----------------------------

> CumulusCI takes an approach to creating and using scratch orgs that aims to make the process easy, portable, and repeatable. An org in CumulusCI's keychain starts out as a named configuration, tailored for a specific purpose within the lifecycle of the project. The scratch org is only actually generated the first time you use the scratch org from the project keychain - and once it's expired or been deleted, a new one can easily be created with the same configuration.

Reviewer: ```suggestion
CumulusCI takes an approach to creating and using scratch orgs that aims to make the process easy, portable, and repeatable. An org in CumulusCI's keychain starts out as a named configuration, tailored for a specific purpose within the lifecycle of the project. The scratch org is only actually generated the first time you use the scratch org - and once it's expired or been deleted, a new one can easily be created with the same configuration.
```

1. `Install the Salesforce CLI <https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_install_cli.htm>`_
2. `Enable Dev Hub in Your Org <https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_enable_devhub.htm>`_
3. `Connect SFDX to Your Dev Hub Org <https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_auth_web_flow.htm>`_ (be sure to use the ``--setdefaultdevhubusername`` option).

If you already have the ``sfdx`` command installed, have connected to your Dev Hub, and have set the ``defaultdevhubusername`` config setting (use ``sfdx force:org:list`` to verify; the default Dev Hub will be listed with a ``(D)``), you're ready to start using CumulusCI with Salesforce CLI.

See Also
^^^^^^^^

* `Learn more about the Salesforce CLI <https://developer.salesforce.com/platform/dx>`_.

Predefined Orgs
---------------

> CumulusCI comes with five org predefined org configurations. Every org's keychain starts with these configurations available and ready to be turned into a live scratch org. You can see the predefined org configurations in your project by running

Reviewer 2: ```suggestion
CumulusCI comes with five predefined org configurations. Every org's keychain starts with these configurations available and ready to be turned into a live scratch org. You can see the predefined org configurations in your project by running
```

> .. code-block: console

Reviewer 2: ```suggestion
.. code-block:: console
```

You can create new orgs in the keychain that inherit their configuration from a built-in org. For example, to create a new org name that uses the same configuration as type ``dev``, you can use the command ::

    $ cci org scratch dev new-org

You can then run ::

    $ cci org info new-org

to create this scratch org, independent of the org ``dev`` but sharing its configuration. You can have as many named orgs as you wish, or none at all: many CumulusCI users work only with the built-in orgs.

Scratch Org Limits
^^^^^^^^^^^^^^^^^^

Each scratch org you create is counted against limits in your Dev Hub. Scratch orgs count against an active scratch org limit, which controls how many orgs you can have at the same time, and a daily scratch org limit, which controls how many total orgs you can create per day.

Scratch org limits are based on your Dev Hub's edition and your Salesforce contract. To review limits and consumption, run the command

> .. code-block: console

Reviewer 2: ```suggestion
.. code-block:: console
```

.. code-block:: console

    $ cci org default dev
    $ cci flow run dev_org

Alternately, you can set a default org when you create a new named configuration:

.. code-block:: console

    $ cci org scratch dev new-org --default

To remove an existing default, run the command

.. code-block:: console

    $ cci org default dev --unset

Configure Predefined Orgs
-------------------------

Projects may choose to customize the set of five configurations available out of the box, and may add further predefined orgs to meet project-specific needs. 

An org configuration has a name, such as ``dev`` or ``qa``, and is defined by options set in ``cumulusci.yml``, plus the contents of a specific ``.json`` scratch org definition file in the ``orgs`` directory. For orgs like ``dev`` and ``qa`` that are predefined for all projects, the configuration is located in the CumulusCI standard library, but can be customized by projects in ``cumulusci.yml``.

Many projects that build managed packages offer a ``dev_namespaced`` org, a developer org that has a namespace. This org is defined like this in ``cumulusci.yml``:

> .. code-block: yaml

Reviewer 2: ```suggestion
.. code-block:: yaml
```

Many projects never need to add a new org definition ``.json`` file, simply modifying the files that are shipped with CumulusCI to add specific features and settings. However, new definitions can be added and referenced in the ``scratch:`` section of ``cumulusci.yml`` to establish org configurations that are completely customized for a project.

Import an Org from the Salesforce CLI
-------------------------------------

CumulusCI can import existing orgs from the Salesforce CLI keychain. To import a scratch org from Salesforce CLI, run

.. code-block:: console

    $ cci org import sfdx_alias cci_alias

For ``sfdx_alias``, you can specify the alias or username of the org in the Salesforce CLI keychain. For ``cci_alias``, provide the name you'd like to use in CumulusCI's keychain.

Note that CumulusCI cannot automatically refresh orgs imported from Salesforce CLI when they expire.

Use a Non-Default Dev Hub
-------------------------

By default, CumulusCI will create scratch orgs using the Dev Hub org that is configured as the ``defaultdevhubusername`` in ``sfdx``. You can switch to a different Dev Hub org within a particular project by configuring the ``devhub`` service:

> .. code-block: console

Reviewer 2: ```suggestion
.. code-block:: console
```

To set up Salesforce DX:

1. `Install Salesforce CLI <https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_install_cli.htm>`_
2. `Enable Dev Hub Features in Your Org <https://developer.salesforce.com/docs/atlas.en-us.228.0.sfdx_dev.meta/sfdx_dev/sfdx_setup_enable_devhub.htm>`_
3. `Connect SFDX to Your Dev Hub Org <https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_auth_web_flow.htm>`_ 

.. important :: Be sure to use the ``--setdefaultdevhubusername`` option!

If you have the ``sfdx`` command installed, are connected to your DevHub, and set the ``defaultdevhubusername`` config setting, you're now ready to use ``cci`` with ``sfdx``.

.. note:: SFDX supports multiple DevHubs, so CumulusCI uses the one set as ``defaultdevhubusername`` when creating scratch orgs. Use ``sfdx force:config:list`` to verify the current value of ``defaultdevhubusername`` on your machine.

The Salesforce Developers site has comprehensive documentation on `Salesforce DX <https://developer.salesforce.com/platform/dx>`_.

Predefined Orgs
---------------

> CumulusCI comes with predefined org configurations. Every org's keychain starts with these configurations ready and available to be turned into a live scratch org.

Reviewer: ```suggestion
CumulusCI comes with predefined org configurations. Every project's keychain starts with these configurations ready and available to be turned into a live scratch org.
```

+-------------+--------------------------+-----------------------+----------+

> |   Org       | Role                     | Configuration File    | Lifespan |

Reviewer: ```suggestion
|   Org       | Role                     | Definition File       | Lifespan |
```

.. code-block:: console

    $ cci org list

If your project has customized org configurations, your listing can include more configurations than shown in the previous table, and your project's versions of the standard configurations can be different.

Create a Scratch Org
--------------------

To create a scratch org from a configuration, use it as the target of a command, task, or flow. CumulusCI automatically initializes orgs when they're first used.

    Example: Create a scratch org from the ``dev`` configuration and review information about the created org.

.. code-block:: console

    $ cci org info dev

..

    When the org is created, it's associated with the name ``dev`` in the CumulusCI keychain and can be used with other commands until it expires. When an org expires or is removed, its associated configuration is left in place, and can be recreated whenever needed.

It's possible to create new orgs in the keychain that inherit their configuration from a built-in org.

    Example: Create a new org that uses the same configuration as type ``dev``, but is still independent of ``dev`` org.  
    
.. code-block ::

    $ cci org scratch dev <org_name>

..

>     Verify that there is now an org with the name of ``<org_name>`` that is associated with the ``orgs/dev.json`` configuration file by running ``cci org list``.     

Reviewer: ```suggestion
    Verify that there is now an org with the name of ``<org_name>`` that is associated with the ``orgs/dev.json`` definition file by running ``cci org list``.     
```

To remove the existing default org:

.. code-block:: console

    $ cci org default dev --unset

Configure Predefined Orgs
-------------------------

Projects can customize the set of configurations available out of the box, and add further predefined orgs to meet project-specific needs. 

An org configuration has a name, such as ``dev`` or ``qa``, and is defined by options set in the ``cumulusci.yml`` file as well as in the contents of a specific ``.json`` scratch org definition file in the ``orgs`` directory. For orgs like ``dev`` and ``qa`` that are predefined for all projects, the configuration is located in the CumulusCI standard library, but can be customized by projects in the ``cumulusci.yml`` file.

When developing a managed package project, it is often useful to test inside of a namespaced scratch org. Many projects configure an org called ``dev_namespaced``, a developer org that has a namespace. This org is defined under the  ``orgs`` section in the ``cumulusci.yml`` file.

.. code-block:: yaml

    orgs:
        scratch:
            dev_namespaced:
                config_file: orgs/dev.json
                days: 7
                namespaced: True

> This org uses the same Salesforce DX configuration file as the ``dev`` org, but has a different configuration in the ``cumulusci.yml`` file, resulting in a different org shape and a different use case. The key facets of the org shape that are defined in the ``cumulusci.yml`` file are whether or not the org has a namespace, and the length of the org's lifespan. 

Reviewer: ```suggestion
This org uses the same scratch org definition file as the ``dev`` org, but has a different configuration in the ``cumulusci.yml`` file, resulting in a different org shape and a different use case. The key facets of the org shape that are defined in the ``cumulusci.yml`` file are whether or not the org has a namespace, and the length of the org's lifespan. 
```

===================

Scratch orgs are temporary Salesforce orgs which can be quickly set up "from scratch," and which last for no more than 30 days. There are several reasons why scratch orgs are encouraged for development and testing over sandboxes or Developer Edition orgs.

* Scratch orgs provide a repeatable starting point without the challenge of managing persistent orgs' state over time.
* Scratch orgs are scalable and ensure that individual, customized environments are available to everyone in the development lifecycle.
* Scratch orgs facilitate a fully source-driven development process built around best practices.

CumulusCI offers tools for working with all types of Salesforce orgs, but provides the most value when working with scratch orgs. CumulusCI automation helps realize the promise of scratch orgs as low cost, repeatable, source-driven environments for every phase of the product lifecycle.

This section focuses on managing scratch orgs in a CumulusCI project. To learn about managing persistent orgs, such as sandboxes, production orgs, and packaging orgs, read `Connect persistent orgs<TODO>`_.

What Is an Org in CumulusCI?
----------------------------

> An org in CumulusCI's keychain starts out as a named configuration, tailored for a specific purpose within the lifecycle of the project (such as development, feature, beta, and so on). CumulusCI creates and uses scratch orgs that are portable and repeatable. In fact, a scratch org is only generated the first time you use the scratch org. When it's expired or been deleted, a new one can easily be created again with the same configuration file.

Reviewer 3: Several times throughout this section "the configuration" is mentioned. I spoke to @Br4nd0R about it, and we settled on "configuration file". I'd like to get a consensus on this before I go changing every single instance of it.

Author: I agree that the terminology is confusing, but I don't think that "configuration" == "configuration file". A scratch org configuration includes both markup in `cumulusci.yml` and in the `orgs/` directory, and goes by a specific name (e.g., `dev`).

Author: This paragraph is incorrect in using "configuration" once and "configuration file" another time. It should be "configuration" both times.

Reviewer 3: @davidmreed Yeah, that's on me. I spoke with @Br4nd0R about clarifying this particular sentence, knowing that there would need to be further discussion on "configuration". I'll make sure it reflects the other instances when we sort out terminology.

Reviewer 3: Gonna close this conversation, but we will have to follow up on whether we say "configuration" or "configuration file" throughout this entire section.

CumulusCI offers tools that make it easy to discover predefined org configurations, create scratch orgs based on those configurations, and define new orgs and new configurations.

> Set Up SFDX 

Reviewer 2: Let's use "Set Up the Salesforce CLI" per discussion in the docs slack group.

> Set Up SFDX 

Reviewer 3: @Br4nd0R Just for later note: should we apply all the edits in this section in the Get Started section as well? Something to consider.

-----------

Scratch orgs in CumulusCI allow teams to be confident that the orgs they develop and test in are as close to their production environments as possible. We recommend working with scratch orgs created by Salesforce DX.

> .. note:: For a detailed introduction to set up Salesforce CLI and Visual Studio Code to work with CumulusCI, we recommend completing the `Build Applications with CumulusCI <https://trailhead.salesforce.com/en/content/learn/trails/build-applications-with-cumulusci>`_ module on Trailhead.

Reviewer 2: I think we can definitely keep this. Just a slight tweak:
```suggestion
.. note:: For a detailed introduction on how to set up Salesforce CLI and Visual Studio Code to work with CumulusCI, we recommend completing the `Build Applications with CumulusCI <https://trailhead.salesforce.com/en/content/learn/trails/build-applications-with-cumulusci>`_ module on Trailhead.
```

To set up Salesforce DX:

1. `Install Salesforce CLI <https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_install_cli.htm>`_
2. `Enable Dev Hub Features in Your Org <https://developer.salesforce.com/docs/atlas.en-us.228.0.sfdx_dev.meta/sfdx_dev/sfdx_setup_enable_devhub.htm>`_
3. `Connect SFDX to Your Dev Hub Org <https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_auth_web_flow.htm>`_ 

.. important :: Be sure to use the ``--setdefaultdevhubusername`` option!

> If you have the ``sfdx`` command installed, are connected to your DevHub, and set the ``defaultdevhubusername`` config setting (use the ``sfdx force:config:list`` command to verify!), you're now ready to use ``cci`` with ``sfdx``.

Reviewer 2: Move the command that instructs how to check into the following admonition.
```suggestion
If you have ``sfdx`` installed, are connected to your DevHub, and set the ``defaultdevhubusername`` config setting,  you're now ready to use ``cci`` with ``sfdx``.
```

> .. note:: SFDX supports multiple DevHubs, so CumulusCI uses the one set as ``defaultdevhubusername`` when creating scratch orgs.

Reviewer 2: ```suggestion
.. note:: SFDX supports multiple DevHubs, so CumulusCI uses the one set as ``defaultdevhubusername`` when creating scratch orgs. Use ``sfdx force:config:list`` to verify the current value of ``defaultdevhubusername`` on your machine.
```

Reviewer 3: Making note in the Doc outline to change this in Get Started as well.

> Learn more about Salesforce DX at https://developer.salesforce.com/platform/dx.

Reviewer 2: I think we should avoid having raw links in the documentation. I can't think of a great way to incorporate this. Maybe something like:
```suggestion
.. tip::
    The Salesforce Developers site has comprehensive documentation regarding `Salesforce DX <https://developer.salesforce.com/platform/dx.>`_.
```

Predefined Orgs
---------------

> CumulusCI comes with predefined org configurations. Every org's keychain starts with these configurations ready and available to be turned into a live scratch org. To see the predefined org configuations in your project:
> 
> .. code-block:: console
> 
>     $ cci org list
> 

Reviewer 2: ```suggestion
CumulusCI comes with predefined org configurations. Every org's keychain starts with these configurations ready and available to be turned into a live scratch org.
```

+-------------+--------------------------+-----------------------+----------+
|   Org       | Role                     | Configuration File    | Lifespan |
+=============+==========================+=======================+==========+
| ``dev``     | Development workflows    | ``orgs/dev.json``     | 7 days   |
+-------------+--------------------------+-----------------------+----------+
| ``qa``      | Testing workflows        | ``orgs/dev.json``     | 7 days   |
+-------------+--------------------------+-----------------------+----------+
| ``feature`` | Continuous Integration   | ``orgs/dev.json``     | 1 day    |
+-------------+--------------------------+-----------------------+----------+
| ``beta``    | | Continuous Integration | ``orgs/beta.json``    | 1 day    |
|             | | Hands-On Testing       |                       |          |
+-------------+--------------------------+-----------------------+----------+
| ``release`` | | Continuous Integration | ``orgs/release.json`` | 1 day    |
|             | | Hands-On Testing       |                       |          |
|             | | Demo Workflow          |                       |          |

> +-------------+--------------------------+-----------------------+----------+

Reviewer 3: The bullet list version of the orgs seemed to be a bit repetitive in its structure, so I thought a table would be even better. (I left the bullet list just below for comparison. Let me know which one you prefer.)

Reviewer 2: I really like how clean this table is. Unfortunately this branch isn't using the updated theme that the final version will have. My vote is still for the table though. I think we should move the command to list the orgs in a project below the table (see edits above and below).

> 

Reviewer 2: ```suggestion
To see the predefined orgs in your project run:

.. code-block:: console

    $ cci org list
```

* ``dev`` is for development workflows and uses the ``orgs/dev.json`` configuration file. It has a seven-day lifespan.
* ``qa`` is for testing workflows. ``qa`` and ``dev`` are the same out of the box, but can be customized to suit the needs of the project. It has a seven-day lifespan.
* ``feature`` is for continuous integration and uses the ``orgs/dev.json`` configuration file. It has a one-day lifespan.
* ``beta`` is for continuous integration or hands-on testing. It uses the ``orgs/beta.json`` configuration. It has a one-day lifespan.

> * ``release`` is for continuous integration, hands-on testing, or demo workflow. It uses the ``orgs/release.json`` configuration. It has a one-day lifespan.

Reviewer 3: To be deleted if table is preferred.

> If your project has customized the available orgs, you'll see more orgs, or differences in configuration, compared to a new project.

Reviewer 3: This kinda goes without saying. Unless I'm missing an insight here, I recommend scrapping it.

Author: Maybe it can be made clearer/

```suggestion
If your project has customized org configurations, your listing may include more configurations than shown above, and your project's versions of the standard configurations may be different.
```

Reviewer 2: I think it's worth keeping. It reconfirms to newer users that the output will likely be different if they're beginning work on an existing project that has been highly configured.

Create a Scratch Org
--------------------

To create a scratch org from a configuration, use it as the target of a command, task, or flow. CumulusCI automatically initializes orgs when they're first used.

    Example: Create a scratch org from the ``dev`` configuration and review information about the created org.

.. code-block:: console

    $ cci org info dev

..

    When the org is created, it's associated with the name ``dev`` in the CumulusCI keychain and can be used with other commands until it expires. When an org expires or is removed, its associated configuration is left in place, and can be recreated whenever needed.

It's possible to create new orgs in the keychain that inherit their configuration from a built-in org.

    Example: Create a new org that uses the same configuration as type ``dev``, but is still independent of ``dev`` org.  
    

> .. code-block ::
> 
>     $ cci org scratch dev <org>
>     $ cci org info <org>

Reviewer 2: The `<org>` variable name just doesn't sit well with me here. We use the word "org" a lot in this section, I'm inclined to go with `<org_name>` especially since we don't mention what the`<org>` parameter does in the command. I know that we haven't been doing that in other places in the doc, but I think it's worth being more precise as it's more informative to end users.

We could also go with a concrete example here:
```suggestion
.. code-block ::

    $ cci org scratch dev myDevOrg

You can verify that there is now an org with the name of ``myDevOrg`` that is associated with the ``orgs/dev.json`` configuration file by running ``cci org list``. 
```

Reviewer 3: Okay. Changing all instances of <org> to <org_name>. But we still need to hash out whether or not we're doing this across the board.

    
You can have as many named orgs as you wish, or none at all. Many CumulusCI users work only with built-in orgs.

Scratch Org Limits
^^^^^^^^^^^^^^^^^^

Each scratch org you create is counted against limits in your DevHub. Scratch orgs count against an *active* scratch org limit, which controls how many orgs can exist at the same time, and a *daily* scratch org limit, which controls how many total orgs can be created per day.

Scratch org limits are based on your DevHub's edition and your Salesforce contract. To review limits and consumption:

.. code-block:: console

    $ sfdx force:limits:api:display -u <username>

``<username>`` is your DevHub username. The limit names are ``ActiveScratchOrgs`` and ``DailyScratchOrgs``.

List Orgs
---------

> When inside a project repository, run ``cci org list`` to see all the orgs you have configured or connected.

Reviewer 3: This subsection is a bit thin. Doesn't necessarily have to be changed, but just wanted eyes on it for confirmation.

Open Orgs in the Browser
---------------------------

> Run ``cci org browser <org>`` to log into any org in the keychain in a new browser tab.

Reviewer 3: This subsection is a bit thin. Doesn't necessarily have to be changed, but just wanted eyes on it for confirmation.

Reviewer 2: Again,I feel like ``<org_name>`` better conveys what that parameter should be.

Delete Scratch Orgs
-------------------

If an org defined in the keychain has created a scratch org, delete the scratch org but leave the configuration to regenerate it in the keychain.

.. code-block:: console

    $ cci org scratch_delete <org>

Using ``scratch_delete`` doesn't remove the org ``<org>`` from your org list.  This default behavior lets you easily recreate scratch orgs from a stored, standardized configuration.

To permanently remove an org from the org list:

.. code-block:: console

    $ cci org remove

To delete a scratch org created from the config:

.. code-block:: console

    $ cci org remove <org>

It's not necessary to explicitly remove or delete expired orgs. CumulusCI recreates an expired org the first time you attempt to use it. To clean up expired orgs from the keychain:

.. code-block:: console

    $ cci org prune

Set a Default Org
-----------------

> When you run a flow or task that performs work on an org, specify the org with an ``--org <org>`` flag.

Reviewer 2: ```suggestion
When you run a flow or task that performs work on an org, specify the org with the ``--org`` option.
```

.. code-block:: console

    $ cci flow run dev_org --org dev

To run many commands against the same org, set a default.

.. code-block:: console

    $ cci org default dev
    $ cci flow run dev_org

> Alternately, set a default org when creating a new named configuration.

Reviewer 2: ```suggestion
Alternately, set a default org when creating a new named configuration by passing the ``--default`` flag.
```

.. code-block:: console

    $ cci org scratch dev <org> --default

> To remove an existing default:

Reviewer 2: ```suggestion
To remove the existing default org:
```

.. code-block:: console

    $ cci org default dev --unset

Configure Predefined Orgs
-------------------------

Projects can customize the set of configurations available out of the box, and add further predefined orgs to meet project-specific needs. 

> An org configuration has a name, such as ``dev`` or ``qa``, and is defined by options set in the ``cumulusci.yml`` file as well as the contents of a specific ``.json`` scratch org definition file in the ``orgs`` directory. For orgs like ``dev`` and ``qa`` that are predefined for all projects, the configuration is located in the CumulusCI standard library, but can be customized by projects in the ``cumulusci.yml`` file.

Reviewer 3: "Specific ``.json`` scratch org definition file" -- Does this file have a name? Should it? Or should we list file path?

Reviewer 2: [Scratch org definition file](https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_scratch_orgs_def_file.htm#!) is the terminology that Salesforce uses. You can have multiple definition files, so we can't provide a specific name. We also indicate that these live in the `orgs` directory. I might add a `/` to that to more clearly indicate it's a directory in a filepath.
```suggestion
An org configuration has a name, such as ``dev`` or ``qa``, and is defined by options set in the ``cumulusci.yml`` file as well as the contents of a specific ``.json`` scratch org definition file in the ``orgs/`` directory. For orgs like ``dev`` and ``qa`` that are predefined for all projects, the configuration is located in the CumulusCI standard library, but can be customized by projects in the ``cumulusci.yml`` file.
```

Reviewer 3: I was just confirming whether or not the specific .json file was easily pinpointed by way of a name or an filepath. No matter then.

> Many projects that build managed packages offer a ``dev_namespaced`` org, a developer org that has a namespace. This org is defined under the  ``orgs`` section in the ``cumulusci.yml`` file.

Reviewer 3: How does this come about? Is ``dev_namespaced`` common practice? Is our audience already familiar with it?

Reviewer 2: We could maybe add an additional sentence in the front here. Something like, "When developing managed package projects, it is often useful to have a 
```suggestion
When developing a managed package project, it is often useful to test inside of a namespaced scratch org. Many projects configure an org called ``dev_namespaced``—a developer org that has a namespace. This org is defined under the  ``orgs`` section in the ``cumulusci.yml`` file.
```

                days: 7
                namespaced: True

This org uses the same Salesforce DX configuration file as the ``dev`` org, but has different configuration in the ``cumulusci.yml`` file, resulting in a different org shape and a different use case. The key facets of the org shape that are defined in the ``cumulusci.yml`` file are whether or not the org has a namespace, and the length of the org's lifespan. 

Org definition files stored in the ``orgs`` directory are configured as specified in the `Salesforce DX Developer Guide <https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_scratch_orgs_def_file.htm>`_.

Many projects never add a new org definition ``.json`` file, and instead add specific features and settings to the files shipped with CumulusCI. However, new definitions can be added and referenced under the ``orgs__scratch`` section of the ``cumulusci.yml`` file to establish org configurations that are completely customized for a project.

Import an Org from the Salesforce CLI
-------------------------------------

CumulusCI can import existing orgs from the Salesforce CLI keychain.

.. code-block:: console

>     $ cci org import sfdx_alias cci_alias

Reviewer 2: ```suggestion
    $ cci org import <sfdx_org_alias> <cci_org_alias>
```

For ``sfdx_alias``, specify the alias or username of the org in the Salesforce CLI keychain. For ``cci_alias``, provide the name to use in CumulusCI's keychain.

> .. note :: CumulusCI cannot automatically refresh orgs imported from Salesforce CLI when they expire.

Reviewer 3: We need more detail about what steps should be taken to refresh/create these imported orgs.

Reviewer 2: ```suggestion
.. important::
    CumulusCI cannot automatically refresh orgs imported from Salesforce CLI when they expire.
```

Use a Non-Default DevHub
-------------------------

By default, CumulusCI creates scratch orgs using the DevHub org configured as the ``defaultdevhubusername`` in ``sfdx``. Switch to a different DevHub org within a project by configuring the ``devhub`` service.

.. code-block:: console

    $ cci service connect devhub --project
    Username: <DevHub username>

>     devhub is now configured for this project.

Reviewer 3: Can someone please confirm that there is an actual period in this last line of code-block text.

Reviewer 2: Confirmed.
